test/t/io/test_metadata.cpp

```diff
@@ -87,6 +91,78 @@ TEST_CASE("Metadata options: timestamp,uid,user") {
 }
 
 TEST_CASE("Metadata options: fail") {
-    REQUIRE_THROWS_AS(osmium::io::detail::metadata_options{"timestamp+foo"}, const std::invalid_argument&);
+    REQUIRE_THROWS_AS(osmium::io::metadata_options{"timestamp+foo"}, const std::invalid_argument&);
+}
+
+TEST_CASE("Metdata options: constructor using OSMObject") {
+    osmium::memory::Buffer buffer{10 * 1000};
+    using namespace osmium::builder::attr;
+
+    SECTION("only version") {
+        const osmium::OSMObject& obj = buffer.get<osmium::OSMObject>(osmium::builder::add_node(buffer, _id(1),
+                _version(2)));
+        osmium::io::metadata_options options = osmium::io::detect_available_metadata(obj);
+        REQUIRE_FALSE(options.user());
+        REQUIRE_FALSE(options.uid());
+        REQUIRE_FALSE(options.changeset());
+        REQUIRE_FALSE(options.timestamp());
+        REQUIRE(options.version());
+    }
+
+    SECTION("full") {
+        const osmium::OSMObject& obj = buffer.get<osmium::OSMObject>(osmium::builder::add_node(buffer, _id(1),
+                _version(2), _cid(30), _user("foo"), _uid(8), _timestamp("2018-01-01T23:00:00Z")));
+        osmium::io::metadata_options options = osmium::io::detect_available_metadata(obj);
+        REQUIRE(options.all());
+    }
+
+    SECTION("changeset+timestamp+version") {
+        const osmium::OSMObject& obj = buffer.get<osmium::OSMObject>(osmium::builder::add_node(buffer, _id(1),
+                _version(2), _cid(30), _timestamp("2018-01-01T23:00:00Z")));
+        osmium::io::metadata_options options = osmium::io::detect_available_metadata(obj);
+        REQUIRE(options.version());
+        REQUIRE(options.timestamp());
+        REQUIRE(options.changeset());
+        REQUIRE_FALSE(options.user());
+        REQUIRE_FALSE(options.uid());
+    }
+}
+
+TEST_CASE("Metdata options: string representation should be valid") {
+
+    SECTION("version+changeset") {
```

Comment:
Can you make these SECTIONs into their own TEST_CASEs? There is no common code in the TEST_CASE, so no need for the hierarchy here.